### PR TITLE
[usb:android.c] Add functionfs support to usb gadget. Fixes: NEMO#811

### PR DESCRIPTION
**Pull request by philippedeswert**
Add functionfs support to the android gadget so that mtp works in
SailfishOS.

Signed-off-by: Philippe De Swert philippe.deswert@jollamobile.com

### CODE REVIEW

**Comment by sledges** [comment]
Thanks for the input, but latest kernel does not boot with your commit (panics at some point and loops on NOPs, as can be told by device heating up), I've not got the stereo->UART cable, so can't give more detailed info unfortunately.

**Comment by philippedeswert** [comment]
Damn, will have to get my device running to check out what goes wrong.

File: drivers/usb/gadget/android.c [hunk 1]
```diff
@@ -2437,15 +2602,37 @@ functions_store(struct device *pdev, struct device_attribute *attr,
 		}
 
 		while (conf_str) {
-			name = strsep(&conf_str, ",");
```
Comment (mlehtima) [inline drivers/usb/gadget/android.c:2440]: I believe this should not be removed. If not removed an adapted version of the patch works also on iyokan.

Comment (philippedeswert) [inline drivers/usb/gadget/android.c:2440]: I believe you're right

**Comment by philippedeswert** [comment]
Missing name variable assignment added, and comments suggest this was the missing bit. So once double-checked this should fix the issue.

**Comment by sledges** [comment]
Tested by @alinelena , works! many thanks! merging

**Comment by simonvanderveldt** [comment]
@philippedeswert Out of interest, why is there new code necessary to make this work? Why is it that Android doesn't need this?

**Comment by philippedeswert** [comment]
@simonvanderveldt : Android uses another system to do the mtp data transfers over the USB port

**Comment by simonvanderveldt** [comment]
@philippedeswert Thx for the info. Does that mean that Android handles mtp in userspace?

**Comment by philippedeswert** [comment]
@simonvanderveldt : Both are handled in userspace but by different software. The transport layer on USB level is different. I don't know however how things are now in android since they have been moving to function fs for things like adb lately.

**Comment by simonvanderveldt** [comment]
@philippedeswert Thanks for the quick responses and the info. I'll try to port your fixes to my kernel as well.

**Comment by philippedeswert** [comment]
@simonvanderveldt : you're welcome. Ask me on IRC if you run into some problem, you know how to find me. nick = phdeswer

**Comment by simonvanderveldt** [comment]
:+1:

**Pull request merged**